File: 프로그래머스/2/49994. 방문 길이/방문 길이.py
def solution(dirs):
    # 방향 설정
    direction = {
        'U': [1, 0],
        'D': [-1, 0],
        'L': [0, -1],
        'R': [0, 1]
    }
    # 방문 여부를 확인할 집합 생성. 집합의 특성을 활용하여 중복을 고려하지 않아도 됨.
    visited = set()

    # 시작 좌표 초기화
    x, y = 0, 0
    
    for d in dirs:
        # 이동할 좌표
        newX = x + direction[d][0]
        newY = y + direction[d][1]
        
        # 유효한 범위 안에 있는지 확인
        if -5 <= newX <= 5 and -5 <= newY <= 5:
            # 시작점과 도착점, 도착점과 시작점 둘다 추가
            # A -> B랑 B -> A는 같은 길을 지난 것이기 때문
            visited.add(str(newX) + str(newY) + str(x) + str(y))
            visited.add(str(x) + str(y) + str(newX) + str(newY))
            x, y = newX, newY
            
    return len(visited) / 2

File: 프로그래머스/2/49994. 방문 길이/방문 길이.js
function solution(dirs) {
    let answer = 0;
    const n = 5;
    const d = {
        'U': [-1, 0],
        'D': [1, 0],
        'L': [0, -1],
        'R': [0, 1]
    };
    const ref = new Map();
    
    let x = 0;
    let y = 0;
    
    dirs.split("").forEach((dir) => {
        const nx = x + d[dir][0];
        const ny = y + d[dir][1];
        
        // 범위 확인
        if (-5 <= nx && nx <= 5 && -5 <= ny && ny <= 5) {
            const path1 = `${x}${y}|${nx}${ny}`;
            const path2 = `${nx}${ny}|${x}${y}`;
            if (!ref.has(path1) && !ref.has(path2)) {
                ref.set(path1, true);
                ref.set(path2, true);
                answer++;
            }
            x = nx;
            y = ny;
        }
    })
    
    return answer;
}